diff --git a/Cargo.toml b/Cargo.toml
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -153,7 +153,7 @@
 version = "2.1"
 
 [dependencies.quick-xml]
-version = "0.39.0"
+version = "0.41.0"
 features = [
     "serialize",
     "overlapped-lists",
diff --git a/Cargo.toml.orig b/Cargo.toml.orig
--- a/Cargo.toml.orig
+++ b/Cargo.toml.orig
@@ -52,7 +52,7 @@
 httparse = { version = "1.8.0", default-features = false, features = ["std"], optional = true }
 hyper = { version = "1.2", default-features = false, optional = true }
 md-5 = { version = "0.10.6", default-features = false, optional = true }
-quick-xml = { version = "0.39.0", features = ["serialize", "overlapped-lists"], optional = true }
+quick-xml = { version = "0.41.0", features = ["serialize", "overlapped-lists"], optional = true }
 rand = { version = "0.10", default-features = false, features = ["std", "std_rng", "thread_rng"], optional = true }
 reqwest = { version = "0.12", default-features = false, features = ["rustls-tls-native-roots", "http2"], optional = true }
 ring = { version = "0.17", default-features = false, features = ["std"], optional = true }
